Source/WebKit/UIProcess/API/glib/WebKitWebExtension.cpp

```diff
@@ -0,0 +1,1156 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "WebKitWebExtension.h"
+
+#include "WebExtension.h"
+#include "WebKitError.h"
+#include "WebKitPrivate.h"
+#include "WebKitWebExtensionInternal.h"
+#include "WebKitWebExtensionMatchPatternPrivate.h"
+#include <WebCore/Icon.h>
+#include <wtf/RefPtr.h>
+#include <wtf/glib/WTFGType.h>
+#include <wtf/text/CString.h>
+
+using namespace WebKit;
+
+/**
+ * WebKitWebExtension:
+ *
+ * Represents a [WebExtension](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions).
+ * 
+ * A #WebKitWebExtension object encapsulates a web extension’s
+ * resources that are defined by a [`manifest.json` file](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json).
+ * 
+ * This class handles the reading and parsing of the manifest file
+ * along with the supporting resources like icons and localizations.
+ *
+ * Since: 2.48
+ */
+
+struct _WebKitWebExtensionPrivate {
+#if ENABLE(WK_WEB_EXTENSIONS)
+    RefPtr<WebExtension> extension;
+    CString defaultLocale;
+    CString displayName;
+    CString displayShortName;
+    CString displayVersion;
+    CString displayDescription;
+    CString displayActionLabel;
+    CString version;
+    GRefPtr<GPtrArray> requestedPermissions;
+    GRefPtr<GPtrArray> optionalPermissions;
+#endif
+};
+
+WEBKIT_DEFINE_FINAL_TYPE(WebKitWebExtension, webkit_web_extension, G_TYPE_OBJECT, GObject)
+
+enum {
+    PROP_0,
+    PROP_MANIFEST_VERSION,
+    PROP_DEFAULT_LOCALE,
+    PROP_DISPLAY_NAME,
+    PROP_DISPLAY_SHORT_NAME,
+    PROP_DISPLAY_VERSION,
+    PROP_DISPLAY_DESCRIPTION,
+    PROP_DISPLAY_ACTION_LABEL,
+    PROP_VERSION,
+    PROP_REQUESTED_PERMISSIONS,
+    PROP_OPTIONAL_PERMISSIONS,
+    PROP_HAS_BACKGROUND_CONTENT,
+    PROP_HAS_PERSISTENT_BACKGROUND_CONTENT,
+    PROP_HAS_INJECTED_CONTENT,
+    PROP_HAS_OPTIONS_PAGE,
+    PROP_HAS_OVERRIDE_NEW_TAB_PAGE,
+    PROP_HAS_COMMANDS,
+    PROP_HAS_CONTENT_MODIFICATION_RULES,
+    N_PROPERTIES,
+};
+
+static GParamSpec* properties[N_PROPERTIES] = { nullptr, };
```

Comment:
Using `std::array` here gives us bounds checking and better compiler diagnostics:

```suggestion
static std::array<GParamSpec*, N_PROPERTIES> properties;
```